src/Algebra/Module/Structures.agda

```diff
@@ -0,0 +1,262 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some algebraic structures defined over some other structure
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Relation.Binary using (Rel; Setoid; IsEquivalence)
+
+module Algebra.Module.Structures
+  {m ℓm} {M : Set m} (_≈ᴹ_ : Rel M ℓm)
+  where
+
+open import Algebra.Bundles
+open import Algebra.Core
+import Algebra.Definitions as Defs
+import Algebra.Module.Definitions.Left as L
+import Algebra.Module.Definitions.Right as R
+open import Algebra.Structures
+open import Data.Product using (_,_; proj₁; proj₂)
+open import Level using (Level; _⊔_)
+
+private
+  variable
+    r ℓr s ℓs : Level
+
+module _ (semiring : Semiring r ℓr) where
+  open Semiring semiring renaming (Carrier to R)
+
+  record IsPreleftSemimodule (+ᴹ : Op₂ M) (*ₗ : Opₗ R M) (0ᴹ : M)
+                             : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    open L R _≈ᴹ_
+    field
+      *ₗ-cong : Congruent _≈_ *ₗ
+      *ₗ-zeroˡ : LeftZero 0# 0ᴹ *ₗ
+      *ₗ-distribʳ : *ₗ DistributesOverʳ _+_ ⟶ +ᴹ
+      *ₗ-identityˡ : LeftIdentity 1# *ₗ
+      *ₗ-assoc : Associative _*_ *ₗ
+      *ₗ-zeroʳ : RightZero 0ᴹ *ₗ
+      *ₗ-distribˡ : *ₗ DistributesOverˡ +ᴹ
+
+  record IsLeftSemimodule (+ᴹ : Op₂ M) (*ₗ : Opₗ R M) (0ᴹ : M)
+                          : Set (r ⊔ m ⊔ ℓr ⊔ ℓm) where
+    open L R _≈ᴹ_
+    field
+      +ᴹ-isCommutativeMonoid : IsCommutativeMonoid _≈ᴹ_ +ᴹ 0ᴹ
+      isPreleftSemimodule : IsPreleftSemimodule +ᴹ *ₗ 0ᴹ
+
+    open IsPreleftSemimodule isPreleftSemimodule public
+
+    open IsCommutativeMonoid +ᴹ-isCommutativeMonoid public
+      using () renaming
+      ( assoc                to +ᴹ-assoc
+      ; comm                 to +ᴹ-comm
+      ; identity             to +ᴹ-identity
+      ; identityʳ            to +ᴹ-identityʳ
+      ; identityˡ            to +ᴹ-identityˡ
+      ; isEquivalence        to M-isEquivalence
```

Comment:
Given the relation is called `≈ᴹ` perhaps `≈ᴹ-isEquivalence` etc. would be more consistent with other renamings?